go/test/endtoend/reparent/emergencyreparent/ers_test.go

```diff
@@ -131,6 +134,105 @@ func TestReparentDownPrimary(t *testing.T) {
 	utils.ResurrectTablet(ctx, t, clusterInstance, tablets[0])
 }
 
+func TestEmergencyReparentWithBlockedPrimary(t *testing.T) {
+	clusterInstance := utils.SetupReparentCluster(t, policy.DurabilitySemiSync)
+	defer utils.TeardownCluster(clusterInstance)
+
+	if clusterInstance.VtTabletMajorVersion < 24 {
+		t.Skip("Skipping test since `DemotePrimary` on earlier versions does not handle blocked primaries correctly")
+	}
+
+	err := clusterInstance.StartVtgate()
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	conn, err := mysql.Connect(ctx, &mysql.ConnParams{
+		Host: clusterInstance.Hostname,
+		Port: clusterInstance.VtgateMySQLPort,
+	})
+	require.NoError(t, err)
+	defer conn.Close()
+
+	_, err = conn.ExecuteFetch("CREATE TABLE test (id INT PRIMARY KEY, msg VARCHAR(64))", 0, false)
+	require.NoError(t, err)
+
+	tablets := clusterInstance.Keyspaces[0].Shards[0].Vttablets
+
+	// Simulate no semi-sync replicas being available by disabling semi-sync on all replicas
+	for _, tablet := range tablets[1:] {
+		utils.RunSQL(ctx, t, "STOP REPLICA", tablet)
+
+		// Disable semi-sync on replicas to simulate blocking
+		semisyncType, err := utils.SemiSyncExtensionLoaded(context.Background(), tablet)
+		require.NoError(t, err)
+		switch semisyncType {
+		case mysql.SemiSyncTypeSource:
+			utils.RunSQL(context.Background(), t, "SET GLOBAL rpl_semi_sync_replica_enabled = false", tablet)
+		case mysql.SemiSyncTypeMaster:
+			utils.RunSQL(context.Background(), t, "SET GLOBAL rpl_semi_sync_slave_enabled = false", tablet)
+		}
+
+		utils.RunSQL(context.Background(), t, "START REPLICA", tablet)
+	}
+
+	// Try performing a write and ensure that it blocks
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// Attempt writing via vtgate against the primary. This should block (because there's no replicas to ack the semi-sync),
+		// and fail once we perform the ERS
+		_, err := conn.ExecuteFetch("insert into test (id, msg) values (1, 'test 1')", 0, false)
+		require.Error(t, err)
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// Just give a tiny bit of time to ensure the write is waiting on the primary,
+		// and that queries from the semi-sync unblocker start piling up
+		time.Sleep(2000 * time.Millisecond)
```

Comment:
This should work, but as word of caution, we've seen flakiness even at such obscene timeouts and even at higher timeouts due to very poor scheduling in CI. If this starts flaking, then my suggestion would be to instead `SHOW PROCESSLIST` until you find your `INSERT`.

Reply:
Hard +1 here 😄

Reply:
Added in https://github.com/vitessio/vitess/pull/18714/commits/88b06ad5376aab2620d69bc937f5d4463b6e3176 and https://github.com/vitessio/vitess/pull/18714/commits/cb19d48de270bb2bcec8905166bc8ffce2a7e578 👍